File: macos/Services/MCP/Tools/Tier2/ExplainQueryTool.swift
// ExplainQueryTool.swift
// Gridex
//
// MCP Tool: Get EXPLAIN plan for a query without executing it.

import Foundation

struct ExplainQueryTool: MCPTool {
    let name = "explain_query"
    let description = "Get EXPLAIN plan for a query without executing it. Helps AI understand query performance."
    let tier = MCPPermissionTier.read

    let inputSchema: [String: Any] = [
        "type": "object",
        "properties": [
            "connection_id": [
                "type": "string",
                "description": "Connection identifier"
            ],
            "sql": [
                "type": "string",
                "description": "SQL query to explain"
            ]
        ],
        "required": ["connection_id", "sql"]
    ]

    func execute(params: JSONValue, context: MCPToolContext) async throws -> MCPToolResult {
        let connectionId = try extractConnectionId(from: params)

        guard let sql = params["sql"]?.stringValue else {
            throw MCPToolError.invalidParameters("sql is required")
        }

        let permission = await context.checkPermission(tier: tier, connectionId: connectionId)
        if let error = permission.errorMessage {
            throw MCPToolError.permissionDenied(error)
        }

        let (adapter, config) = try await context.getAdapter(for: connectionId)

        // Build EXPLAIN query based on database type
        let explainSQL: String
        switch config.databaseType {
        case .postgresql:
            explainSQL = "EXPLAIN (ANALYZE false, COSTS true, FORMAT TEXT) \(sql)"
        case .mysql:
            explainSQL = "EXPLAIN \(sql)"
        case .sqlite:
            explainSQL = "EXPLAIN QUERY PLAN \(sql)"
        case .mssql:
            // SQL Server uses SET SHOWPLAN_TEXT or estimated plan
            explainSQL = "SET SHOWPLAN_TEXT ON; \(sql); SET SHOWPLAN_TEXT OFF"
        case .mongodb, .redis:
            return MCPToolResult(text: "EXPLAIN is not supported for \(config.databaseType.displayName) connections.", isError: true)
        }

        let result = try await adapter.executeRaw(sql: explainSQL)

        // Format the explain output
        var output = "Query Plan for: \(sql)\n\n"

        if result.rows.isEmpty {
            output += "No plan information available."
        } else {
            for row in result.rows {
                let line = row.map(\.displayString).joined(separator: " | ")
                output += line + "\n"
            }
        }

        return MCPToolResult(text: output)
    }
}

File: windows/Gridex/Services/MCP/Tools/Tier2/ExplainQueryTool.cpp
//
// ExplainQueryTool.cpp
//

#include "ExplainQueryTool.h"
#include "../MCPToolHelpers.h"

namespace DBModels
{
    MCPToolResult ExplainQueryTool::execute(const nlohmann::json& params, MCPToolContext& ctx)
    {
        const auto connId = MCPTool::extractConnectionId(params);

        if (!params.contains("sql") || !params["sql"].is_string())
            throw MCPToolError::invalidParameters("sql is required");
        const auto sqlUtf8 = params["sql"].get<std::string>();

        const auto perm = ctx.checkPermission(tier(), connId);
        if (const auto* msg = perm.errorMessage())
            throw MCPToolError::permissionDenied(*msg);

        auto [adapter, config] = ctx.getAdapter(connId);
        const std::wstring sqlW = MCPToolHelpers::fromUtf8(sqlUtf8);

        std::wstring explainSQL;
        switch (config.databaseType)
        {
            case DatabaseType::PostgreSQL:
                explainSQL = L"EXPLAIN (ANALYZE false, COSTS true, FORMAT TEXT) " + sqlW;
                break;
            case DatabaseType::MySQL:
                explainSQL = L"EXPLAIN " + sqlW;
                break;
            case DatabaseType::SQLite:
                explainSQL = L"EXPLAIN QUERY PLAN " + sqlW;
                break;
            case DatabaseType::MSSQLServer:
                // SET SHOWPLAN_TEXT is a session flag — bracket the
                // query with it. Multi-statement; the adapter passes
                // this through to the driver unchanged.
                explainSQL = L"SET SHOWPLAN_TEXT ON; " + sqlW + L"; SET SHOWPLAN_TEXT OFF";
                break;
            default:
                return MCPToolResult::error(
                    "EXPLAIN is not supported for " +
                    MCPToolHelpers::toUtf8(DatabaseTypeDisplayName(config.databaseType)) +
                    " connections.");
        }

        QueryResult r;
        try { r = adapter->execute(explainSQL); }
        catch (const std::exception& e)
        { throw MCPToolError::queryFailed(e.what()); }
        if (!r.success) throw MCPToolError::queryFailed(MCPToolHelpers::toUtf8(r.error));

        std::string out = "Query Plan for: " + sqlUtf8 + "\n\n";
        if (r.rows.empty())
        {
            out += "No plan information available.";
        }
        else
        {
            for (const auto& row : r.rows)
            {
                bool first = true;
                for (const auto& cname : r.columnNames)
                {
                    auto it = row.find(cname);
                    if (it == row.end()) continue;
                    if (!first) out += " | ";
                    first = false;
                    out += MCPToolHelpers::toUtf8(it->second);
                }
                out += "\n";
            }
        }
        return MCPToolResult::text(out);
    }
}
